### PR TITLE
feat: add conditions to flow topology

### PR DESCRIPTION
**Pull request by JoanFM**

### CODE REVIEW

**Review by bwanglzu — COMMENTED** [review]
how do you think replace:
 `graph_condition` -> `graph_mask` and `condition` -> `mask`, let's say you want to only keep `Document` with `key=5`, I'll do `mask key!=5`. 
 I feel this way is more natural.

**Comment by JoanFM** [comment]
> how do you think replace: `graph_condition` -> `graph_mask` and `condition` -> `mask`, let's say you want to only keep `Document` with `key=5`, I'll do `mask key!=5`. I feel this way is more natural.

Let's see how the `Query Language` in `DocumentArray` ends up being and we can decide? I think this would heavily depend on how it is implemented there. What do u think @bwanglzu ?

**Comment by JohannesMessner** [comment]
The actual code should be ready, so feel free to review.
However, it is not ready to merge yet, since:
- QL feature is not really released in DocArray, so can't link to their docs
- Tutorial is referenced in the docs but not done yet

File: docs/fundamentals/flow/create-flow.md [hunk 1]
```diff
@@ -453,9 +453,172 @@ This will get you the following output:
...
+[]
+```
+````
+
+This feature is useful to prevent some specialized Executors from processing certain Documents.
+It can also be used to build *switch-like nodes*, where some Documents pass through one parallel branch of the Flow,
+while other Documents pass through a different branch.
+
+Also note that whenever a Document does not satisfy the condition of an Executor, it will not even be sent to that Executor.
+This means that you can not only use this feature to build complex logic, but also to minimize your networking overhead.
```
Comment (JoanFM) [inline docs/fundamentals/flow/create-flow.md:616]: this will not remain true right?

File: jina/serve/runtimes/gateway/request_handling.py [hunk 2]
```diff
@@ -25,6 +26,9 @@ def _handle_request(request: 'Request') -> 'asyncio.Future':
 
         request_graph = copy.deepcopy(graph)
         # important that the gateway needs to have an instance of the graph per request
+        request_doc_ids = request.data.docs[
```
Comment (JoanFM) [inline jina/serve/runtimes/gateway/request_handling.py:29]: We need to detect if there is a condition in the `graph` before doing it.

Doing serialization is rather `expensive` and I would rather avoid this if not needed

**Review by JoanFM — COMMENTED** [review]
Just a minor thing

File: docs/fundamentals/flow/create-flow.md (hunk 1 above)
Comment (JohannesMessner) [reply docs/fundamentals/flow/create-flow.md:617]: Added a clarification

**Pull request closed without merging**